README.md
## connector-ipfs (uplink v1.0.5)

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/ac7bbc539c0a45a5a2140b3e6b7c823d)](https://app.codacy.com/gh/storj-thirdparty/connector-IPFS?utm_source=github.com&utm_medium=referral&utm_content=storj-thirdparty/connector-IPFS&utm_campaign=Badge_Grade_Dashboard)
[![Go Report Card](https://goreportcard.com/badge/github.com/storj-thirdparty/connector-ipfs)](https://goreportcard.com/report/github.com/storj-thirdparty/connector-ipfs)
![Cloud Build](https://storage.googleapis.com/storj-utropic-services-badges/builds/connector-ipfs/branches/master.svg)

## Overview

The IPFS Connector connects to an IPFS server, takes a backup of the specified files and uploads the backup data on Storj network.

```bash
Usage:
  connector-ipfs [command] <flags>

Available Commands:
  help        Help about any command
  store       Command to upload data to a Storj V3 network.
  version     Prints the version of the tool

```

`store` - Connect to the specified IPFS (default: `ipfs_property.json`). Back-up of the IPFS is generated using tooling provided by IPFS and then uploaded to the Storj network.  Connect to a Storj v3 network using the access specified in the Storj configuration file (default: `storj_config.json`).

Sample configuration files are provided in the `./config` folder.

## Requirements and Install

To build from scratch, [install the latest Go](https://golang.org/doc/install#install).

> Note: Ensure go modules are enabled (GO111MODULE=on)

#### Option #1: clone this repo (most common)

To clone the repo

```
git clone https://github.com/storj-thirdparty/connector-ipfs.git
```

Then, build the project using the following:

```
cd connector-ipfs
go build
```

#### Option #2:  ``go get`` into your gopath

 To download the project inside your GOPATH use the following command:

```
go get github.com/storj-thirdparty/connector-ipfs
```

## Connect to IPFS Server

Make sure you are connected to IPFS server. If not, run the ipfs daemon in another `terminal` to join your node to the public network:

```
$ ipfs daemon
```

## Run (short version)

Once you have built the project run the following commands as per your requirement:

##### Get help

```
$ ./connector-ipfs --help
```

##### Check version

```
$ ./connector-ipfs --version
```

##### Create backup from ipfs and upload to Storj

```
$ ./connector-ipfs store
```

## Documentation

* Access documentation on local system :
  1) Install [docsify](https://www.npmjs.com/package/docsify-cli)
  2) Run the following command at the root directory of the cloned project.
  ```
  $ docsify serve docs
  ```
* For more information on runtime flags, configuration, testing, and diagrams, check out the [Detail](//github.com/storj-thirdparty/wiki/Detail) or jump to:
  * [Config Files](//github.com/storj-thirdparty/connector-ipfs/wiki/#config-files)
  * [Run (long version)](//github.com/storj-thirdparty/connector-ipfs/wiki/#run)
  * [Testing](//github.com/storj-thirdparty/connector-ipfs/wiki/#testing)
  * [Flow Diagram](//github.com/storj-thirdparty/connector-ipfs/wiki/#flow-diagram)
  * [Video](//github.com/storj-thirdparty/connector-ipfs/docs/videos)


CHANGELOG.md
# connector-ipfs Changelog

## [1.0.5] - 17-09-2020
### Changelog:
*Add video section in ReadME.md


## [1.0.5] - 27-08-2020
### Changelog:
* Added CI-CD pipeline.


## [1.0.5] - 18-08-2020
### Changelog:
* Resolved upload path issue.


## [1.0.5] - 27-05-2020
### Changelog:
* Added cobra cli for user interface.
* Restructured project based on the requirements for cobra cli.
* Changed arguments to optional flags.
* Added `--accesskey` flag and removed `key`, `test`, `parse` and `debug` flags.
* Added `--storj` flag to set storj config file path and `--ipfs` to set ipfs config file path.
* Fixed bugs


## [1.0.7] - 04-12-2019
### Changelog:
* Added Macroon functionality.
* Added option to access storj using Serialized Scope Key. 
* Added keyword `key` to access storj using API key rather than Serialized Scope Key (defalt).
* Added keyword `restrict` to apply restrictions on API key and provide shareable Serialized Scope Key for users.


## [1.0.6] - 29-11-2019
### Changelog:
* Optimized storj-download.
* Exception handling in storj connection.
* Made changes according to updated libuplink package.


## [1.0.5] - 25-11-2019
### Changelog:
* Removed parse and parse debug commands.
* Error handling for various events.


## [1.0.4] - 22-11-2019
### Changelog:
* Added download debug command.
* Updated and optimized test command.
* Optimized upload process by using io.reader instead to bytes.reader.
* The program shows error if daemon is not running.
* fileHash has renamed to shareableHash in ipfs_download.json.
* Retry mechanism is added for upload function.
* Insufficent memory(system freezing) issue when uploading large data resolved.


## [1.0.3] - 15-11-2019
### Changelog:
* Made storj-ipfs-connector able to upload large files.
* Changes made in json fies.


## [1.0.2] - 12-11-2019
### Changelog:
* Made ipfsHostName and PortNumber configurable.
* Changes made according to updated cli package.


## [1.0.1] - 08-11-2019
### Changelog:
* Added download
* Added upload debug
* Added test
* Added storj-ipfs as module
* Changed main project folder to storj-ipfs-connector


## [1.0.0] - 05-11-2019


cmd/main_test.go
package cmd_test

import (
	"context"
	"fmt"
	"log"
	"os"

	"testing"

	"github.com/storj-thirdparty/connector-ipfs/cmd"
)

func TestMongoStore(t *testing.T) {

	storjConfig := cmd.LoadStorjConfiguration("../config/storj_config_test.json")
	_, project := cmd.ConnectToStorj(storjConfig, false)

	fileReader, err := os.Open("../testFile.txt")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Initiating back-up.\n")
	cmd.UploadData(project, storjConfig, "testFile.txt", fileReader)
	fmt.Printf("Back-up complete.\n\n")

	fmt.Printf("\nDeleting the test back-up.\n")
	ctx := context.Background()
	backups := project.ListObjects(ctx, storjConfig.Bucket, nil)
	// Loop to find the latest back-up of all the back-ups.
	for backups.Next() {
		item := backups.Item()
		_, err := project.DeleteObject(ctx, storjConfig.Bucket, item.Key)
		if err != nil {
			log.Fatal(err)
		}
	}
	fmt.Printf("Deleted the test back-up.\n")
}


test/doc-check.sh
## verify documentation is present. 
## cd is /workspace
HOME_MD="./README.assets/HOME.md"
README_MD="./README.md"
if [ -f $HOME_MD ] && [ -f $README_MD ]; then
    echo "doco check PASSED"
    exit 0
fi
echo "doco check FAILED.  Missing either" $README_MD "or" $HOME_MD
exit 1


cmd/ipfs.go
package cmd

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"

	shell "github.com/ipfs/go-ipfs-api"
)

// ConfigIpfs defines the variables and types for login.
type ConfigIpfs struct {
	HostName string `json:"hostName"`
	Port     string `json:"port"`
	Hash     string `json:"hash"`
}

// LoadIpfsProperty reads and parses the JSON file
// that contains a IPFS instance's properties
// and returns all the properties as an object.
func LoadIpfsProperty(fullFileName string) ConfigIpfs {

	var configIpfs ConfigIpfs

	// Open and read the file.
	fileHandle, err := os.Open(filepath.Clean(fullFileName))
	if err != nil {
		log.Fatal(err)
	}

	jsonParser := json.NewDecoder(fileHandle)
	if err = jsonParser.Decode(&configIpfs); err != nil {
		log.Fatal(err)
	}

	err = fileHandle.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Display Information about IPFS Instance.
	fmt.Println("Read IPFS configuration from the ", fullFileName, " file")
	fmt.Println("Host Name\t: ", configIpfs.HostName)
	fmt.Println("Port\t\t: ", configIpfs.Port)
	fmt.Println("Hash\t\t:", configIpfs.Hash)

	return configIpfs
}

// ConnectToIpfs will connect to an IPFS instance
// based on the read property from an external file.
// It returns a reference to shell with IPFS instance's information.
func ConnectToIpfs(configIpfs ConfigIpfs) *shell.Shell {

	fmt.Println("Connecting to IPFS...")

	if configIpfs.HostName == "ipfsHostName" || configIpfs.HostName == "" {
		err1 := errors.New("Invalid HostName")
		log.Fatal("Invalid Hostname error : ", err1)
	}
	// Connect IPFS deamon to IPFS node.
	sh := shell.NewShell(configIpfs.HostName + ":" + configIpfs.Port)
	_, _, errVer := sh.Version()
	if errVer != nil {
		err1 := errors.New("Could not find Daemon running")
		log.Fatal("Daemon error : ", err1)
	}

	return sh
}

// AllHash is a list to store hashes of all files to store from IPFS.
var AllHash []string

// AllFilesWithPaths is a list to store complete path of all the hash in the IPFS.
// It will be used for direct transfer of files from IPFS to Storj.
var AllFilesWithPaths []string

// GetHash retrieve the hash with the exact
// file structure from IPFS to the System
func GetHash(sh *shell.Shell, hash string, path string) {

	files, err := sh.FileList(hash)
	if err != nil {
		log.Fatal(err)
	}
	if files.Type == "Directory" {
		if len(files.Links) > 0 {
			for i := 0; i < len(files.Links); i++ {
				if files.Links[i].Type == "Directory" {
					GetHash(sh, files.Links[i].Hash, path+"/"+files.Links[i].Hash)
				} else {
					AllHash = append(AllHash, files.Links[i].Hash)
					AllFilesWithPaths = append(AllFilesWithPaths, path+"/"+files.Links[i].Hash)
				}
			}
		}
	} else {
		AllHash = append(AllHash, hash)
		if AllFilesWithPaths != nil {
			AllFilesWithPaths = append(AllFilesWithPaths, path+"/"+hash)
		} else {
			AllFilesWithPaths = append(AllFilesWithPaths, hash)
		}
	}
}

// GetReader returns a Reader of corresponding file whose path is specified.
// io.ReadCloser type of object returned is used to perform transfer of file to Storj.
func GetReader(sh *shell.Shell, path string) io.ReadCloser {

	ipfsReader, err := sh.Cat(path)
	if err != nil {
		log.Fatal("Read file error: ", err)
	}

	return ipfsReader
}


cmd/store.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
)

// storeCmd represents the store command.
var storeCmd = &cobra.Command{
	Use:   "store",
	Short: "Command to upload data to storj V3 network.",
	Long:  `Command to connect to desired IPFS account and back-up the complete data to given Storj Bucket.`,
	Run:   ipfsStore,
}

func init() {

	// Setup the store command with its flags.
	rootCmd.AddCommand(storeCmd)
	var defaultIpfsFile string
	var defaultStorjFile string
	storeCmd.Flags().BoolP("accesskey", "a", false, "Connect to storj using access key(default connection method is by using API Key).")
	storeCmd.Flags().BoolP("share", "s", false, "For generating share access of the uploaded backup file.")
	storeCmd.Flags().StringVarP(&defaultIpfsFile, "ipfs", "i", "././config/ipfs_property.json", "full filepath contaning IPFS configuration.")
	storeCmd.Flags().StringVarP(&defaultStorjFile, "storj", "u", "././config/storj_config.json", "full filepath contaning storj V3 configuration.")
}

func ipfsStore(cmd *cobra.Command, args []string) {

	// Process arguments from the CLI.
	ipfsConfigfilePath, _ := cmd.Flags().GetString("ipfs")
	fullFileNameStorj, _ := cmd.Flags().GetString("storj")
	useAccessKey, _ := cmd.Flags().GetBool("accesskey")
	useAccessShare, _ := cmd.Flags().GetBool("share")

	// Read IPFS instance's configurations from an external file and create an IPFS configuration object.
	configIpfs := LoadIpfsProperty(ipfsConfigfilePath)

	// Read storj network configurations from and external file and create a storj configuration object.
	storjConfig := LoadStorjConfiguration(fullFileNameStorj)

	// Connect to storj network using the specified credentials.
	access, project := ConnectToStorj(storjConfig, useAccessKey)

	// Connect to IPFS using the specified credentials
	ipfsShell := ConnectToIpfs(configIpfs)

	// Create the slices of hash names and file hash paths of files to be uploaded.
	GetHash(ipfsShell, configIpfs.Hash, configIpfs.Hash)

	fmt.Printf("\nInitiating back-up.\n")
	// Fetch all hash from IPFS instance and simultaneously store them into desired Storj bucket.
	for i := 0; i < len(AllHash); i++ {
		file := AllFilesWithPaths[i]
		ipfsReader := GetReader(ipfsShell, AllHash[i])
		UploadData(project, storjConfig, file, ipfsReader)
	}
	fmt.Printf("\nBack-up complete.\n\n")

	// Create restricted shareable serialized access if share is provided as argument.
	if useAccessShare {
		ShareAccess(access, storjConfig)
	}
}


cmd/storj.go
package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"storj.io/uplink"
)

// ConfigStorj depicts keys to search for within the stroj_config.json file.
type ConfigStorj struct {
	APIKey               string `json:"apikey"`
	Satellite            string `json:"satellite"`
	Bucket               string `json:"bucket"`
	UploadPath           string `json:"uploadPath"`
	EncryptionPassphrase string `json:"encryptionpassphrase"`
	SerializedAccess     string `json:"serializedAccess"`
	AllowDownload        string `json:"allowDownload"`
	AllowUpload          string `json:"allowUpload"`
	AllowList            string `json:"allowList"`
	AllowDelete          string `json:"allowDelete"`
	NotBefore            string `json:"notBefore"`
	NotAfter             string `json:"notAfter"`
}

// LoadStorjConfiguration reads and parses the JSON file that contain Storj configuration information.
func LoadStorjConfiguration(fullFileName string) ConfigStorj {

	var configStorj ConfigStorj
	fileHandle, err := os.Open(filepath.Clean(fullFileName))
	if err != nil {
		log.Fatal("Could not load storj config file: ", err)
	}

	jsonParser := json.NewDecoder(fileHandle)
	if err = jsonParser.Decode(&configStorj); err != nil {
		log.Fatal(err)
	}

	// Close the file handle after reading from it.
	if err = fileHandle.Close(); err != nil {
		log.Fatal(err)
	}

	// Display storj configuration read from file.
	fmt.Println("\nRead Storj configuration from the ", fullFileName, " file")
	fmt.Println("API Key\t\t: ", configStorj.APIKey)
	fmt.Println("Satellite	: ", configStorj.Satellite)
	fmt.Println("Bucket		: ", configStorj.Bucket)

	// Convert the upload path to standard form.
	if configStorj.UploadPath != "" {
		if configStorj.UploadPath == "/" {
			configStorj.UploadPath = ""
		} else {
			checkSlash := configStorj.UploadPath[len(configStorj.UploadPath)-1:]
			if checkSlash != "/" {
				configStorj.UploadPath = configStorj.UploadPath + "/"
			}
		}
	}

	fmt.Println("Upload Path\t: ", configStorj.UploadPath)
	fmt.Println("Serialized Access Key\t: ", configStorj.SerializedAccess)
	return configStorj
}

// ShareAccess generates and prints the shareable serialized access
// as per the restrictions provided by the user.
func ShareAccess(access *uplink.Access, configStorj ConfigStorj) {

	allowDownload, _ := strconv.ParseBool(configStorj.AllowDownload)
	allowUpload, _ := strconv.ParseBool(configStorj.AllowUpload)
	allowList, _ := strconv.ParseBool(configStorj.AllowList)
	allowDelete, _ := strconv.ParseBool(configStorj.AllowDelete)
	notBefore, _ := time.Parse("2006-01-02_15:04:05", configStorj.NotBefore)
	notAfter, _ := time.Parse("2006-01-02_15:04:05", configStorj.NotAfter)

	permission := uplink.Permission{
		AllowDownload: allowDownload,
		AllowUpload:   allowUpload,
		AllowList:     allowList,
		AllowDelete:   allowDelete,
		NotBefore:     notBefore,
		NotAfter:      notAfter,
	}

	// Create shared access.
	sharedAccess, err := access.Share(permission)
	if err != nil {
		log.Fatal("Could not generate shared access: ", err)
	}

	// Generate restricted serialized access.
	serializedAccess, err := sharedAccess.Serialize()
	if err != nil {
		log.Fatal("Could not serialize shared access: ", err)
	}
	fmt.Println("Shareable serialized access: ", serializedAccess)
}

// ConnectToStorj reads Storj configuration from given file
// and connects to the desired Storj network.
// It then reads data property from an external file.
func ConnectToStorj(configStorj ConfigStorj, accesskey bool) (*uplink.Access, *uplink.Project) {

	var access *uplink.Access
	var cfg uplink.Config

	// Configure the UserAgent.
	cfg.UserAgent = "IPFS"
	ctx := context.Background()
	var err error

	if accesskey {
		fmt.Println("\nConnecting to Storj network using Serialized access.")
		// Generate access handle using serialized access.
		access, err = uplink.ParseAccess(configStorj.SerializedAccess)
		if err != nil {
			log.Fatal(err)
		}
	} else {
		fmt.Println("\nConnecting to Storj network.")
		// Generate access handle using API key, satellite url and encryption passphrase.
		access, err = cfg.RequestAccessWithPassphrase(ctx, configStorj.Satellite, configStorj.APIKey, configStorj.EncryptionPassphrase)
		if err != nil {
			log.Fatal(err)
		}
	}

	// Open a new porject.
	project, err := cfg.OpenProject(ctx, access)
	if err != nil {
		log.Fatal(err)
	}
	defer project.Close()

	// Ensure the desired Bucket within the Project.
	_, err = project.EnsureBucket(ctx, configStorj.Bucket)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Successfully connected to Storj network.")
	return access, project
}

// UploadData uploads the backup file to storj network.
func UploadData(project *uplink.Project, configStorj ConfigStorj, uploadFileName string, reader io.Reader) {

	ctx := context.Background()

	// Create an upload handle.
	upload, err := project.UploadObject(ctx, configStorj.Bucket, configStorj.UploadPath+uploadFileName, nil)
	if err != nil {
		log.Fatal("Could not initiate upload : ", err)
	}
	fmt.Printf("\nUploading %s to %s.", configStorj.UploadPath+uploadFileName, configStorj.Bucket)

	// Upload data on storj.
	_, err = io.Copy(upload, reader)
	if err != nil {
		abortErr := upload.Abort()
		log.Fatal("Could not upload data to storj: ", err, abortErr)
	}

	// Commit the upload after copying the complete content of the backup file to upload object.
	fmt.Println("\nPlease wait while the upload is being committed to Storj.")
	err = upload.Commit()
	if err != nil {
		log.Fatal("Could not commit object upload : ", err)
	}
}
